kep/640-error-pods-rescheduling/README.md

```diff
@@ -0,0 +1,166 @@
+If the current scheduling count does not exceed the user-configured `scheduling-retries` (maximum rescheduling retry count), 
+it increments the current scheduling count by 1 and records the current node where the pod is located in the scheduled nodes list (eg: defined as `deployScheduledHosts`). 
+Then, rescheduler-controller deletes the abnormal pod, and this pod's associated workload resource will create a new pod automatically. 
+Rescheduler-controller registers a MutateWebHook in the kube-apiserver, this MutateWebHook will patch scheduled nodes list info to this newly created pod's annotations (defined as `sigs.k8s.io/scheduled-hosts`) when this newly created pod comes to kube-apiserver admission point.  
+Podrescheduling is the scheduler-plugins that implement the PreFilter extension point of kube-scheduler to prefilter out the scheduled nodes list recorded in this pod annotation and schedule this pod to other nodes.
+
+![design](./design.png)
+
+### Rescheduler-controller
+
+The rescheduler-controller is comprised of two components, as illustrated in the above diagram: ListAndDeleteFunc and MutateWebHook.
+
+The ListAndDeleteFunc serves as a periodic monitoring program responsible for listing pods in abnormal states within the cluster. 
+It is also in charge of deleting these abnormal pods and updating annotations in the metadata of these pods' workload resources to record the scheduled nodes list. 
+The ListAndDeleteFunc triggers pod rescheduling by deleting pods. Additionally, for pods without an owner reference definition, the ListAndDeleteFunc maintains a map (for updating pod's metada annotations) in memory to store scheduled hosts, allowing for direct code-based resubmission of these pods to the cluster.
+
+The MutateWebHook registers with the kube-apiserver and intercepts incoming pods within the cluster. Subsequently, the MutateWebHook patches the list of scheduled nodes for these pods, which is retrieved from the annotations of the pods' associated workload resources, into the metada annotations of the pods themselves.
+
+#### Scheduled nodes list and current rescheduling times definition
+
+for pods' workload resources，like Deployment
+
+```golang
+type DeployInfo struct {
+	CurrentReschedulingTimes int `json:"currentReschedulingTimes, omitempty"`
+	DeployScheduledHosts []string `json:"deployScheduledHosts, omitempty"`
+}
+```
+for pods
+
+```golang
+type PodInfo struct {
+	CurrentReschedulingTimes int `json:"currentReschedulingTimes, omitempty"`
+	PodScheduledHosts []string `json:"podScheduledHosts, omitempty"`
+}
+```
+
+### Podrescheduling
+
+Podrescheduling implements the PreFilter interface of the scheduler framework. It extracts the list of nodes already scheduled from the "`sigs.k8s.io/scheduled-hosts`" field in the pod's metadata annotations. 
+Then, it filters out these already scheduled nodes and returns a list of nodes in the cluster where the pod has not yet been run.
```

Comment:
Mostly curious, why in `PreFilter` and not in `Filter`?

Reply:
In `PreFilter` phase, we can preFilter many nodes at a time, because all the already scheduled nodes are recorded in `sigs.k8s.io/scheduled-hosts`.
But in `Filter` phase, we can only filter one node at a time. So we implement Podrescheduling in `PreFilter` phase.